APIKit/Request.swift

```diff
@@ -1,9 +1,95 @@
 import Foundation
+import Result
 
+/// Request protocol represents a request for Web API.
+/// Following 5 items must be implemented.
+/// - typealias Response
+/// - var baseURL: NSURL
+/// - var method: Method
+/// - var path: String
+/// - func responseFromObject(object: AnyObject, URLResponse: NSHTTPURLResponse) throws -> Response
 public protocol Request {
-    typealias Response: Any
-    
-    var URLRequest: NSURLRequest? { get }
-    
-    static func responseFromObject(object: AnyObject) -> Response?
+    /// Type represents a model object
+    typealias Response
+
+    /// Configurations of request
+    var baseURL: NSURL { get }
+    var method: HTTPMethod { get }
+    var path: String { get }
+    var parameters: [String: AnyObject] { get }
+
+    /// You can add any configurations here
+    func configureURLRequest(URLRequest: NSMutableURLRequest) throws -> NSMutableURLRequest
+
+    /// Set of status code that indicates success.
+    /// `responseFromObject(_:URLResponse:)` will be called if this contains NSHTTPURLResponse.statusCode.
+    /// Otherwise, `errorFromObject(_:URLResponse:)` will be called.
+    var acceptableStatusCodes: Set<Int> { get }
+
+    /// An object that builds body of HTTP request.
+    var requestBodyBuilder: RequestBodyBuilder { get }
+
+    /// An object that parses body of HTTP response.
+    var responseBodyParser: ResponseBodyParser { get }
+
+    /// Build `Response` instance from raw response object.
+    /// This method will be called if `acceptableStatusCode` contains status code of NSHTTPURLResponse.
+    func responseFromObject(object: AnyObject, URLResponse: NSHTTPURLResponse) throws -> Response
+
+    /// Build `ErrorType` instance from raw response object.
+    /// This method will be called if `acceptableStatusCode` does not contain status code of NSHTTPURLResponse.
+    func errorFromObject(object: AnyObject, URLResponse: NSHTTPURLResponse) throws -> ErrorType
```

Comment:
I'm a bit skeptical about throwing error here. Its return type is an `ErrorType`, so returning some errors instead of throwing them is sufficient or not?